File: basic/09-bfs/09-bfs.js
class Node {
  constructor(data) {
      this.data = data;
      this.left = null;
      this.right = null;
  }
}

function bfs(root) {
  if (root === null) return;

  const queue = [root];
  while (queue.length > 0) {
      const current = queue.shift();
      console.log(current.data);

      if (current.left) queue.push(current.left);
      if (current.right) queue.push(current.right);
  }
}

// 示例
const root = new Node(1);
root.left = new Node(2);
root.right = new Node(3);
root.left.left = new Node(4);
root.left.right = new Node(5);

console.log("BFS遍历结果:");
bfs(root);

File: basic/graph/bfs/09-bfs.rb
class Node
  attr_accessor :data, :left, :right

  def initialize(data)
      @data = data
      @left = nil
      @right = nil
  end
end

def bfs(root)
  return if root.nil?

  queue = [root]
  until queue.empty?
      current = queue.shift
      print "#{current.data} "

      queue.push(current.left) if current.left
      queue.push(current.right) if current.right
  end
end

# 示例
root = Node.new(1)
root.left = Node.new(2)
root.right = Node.new(3)
root.left.left = Node.new(4)
root.left.right = Node.new(5)

print "BFS遍历结果: "
bfs(root)
puts
